persist/sqlite/wallet.go

```diff
@@ -641,14 +641,31 @@ func fillElementProofs(tx *txn, indices []uint64) (proofs [][]types.Hash256, _ e
 }
 
 func getWalletEvents(tx *txn, id wallet.ID, offset, limit int) (events []wallet.Event, eventIDs []int64, err error) {
-	const query = `SELECT ev.id, ev.event_id, ev.maturity_height, ev.date_created, ci.height, ci.block_id, ev.event_type, ev.event_data
+	// the events query can be slow in full index mode for wallets with no
+	// events. Check if the wallet has events first.
+	const hasEventsQuery = `SELECT EXISTS (
+  SELECT 1
+  FROM event_addresses ea
+  INNER JOIN wallet_addresses wa ON ea.address_id = wa.address_id
+  WHERE wa.wallet_id=$1
+) AS has_events;`
+	var hasEvents bool
+	if err := tx.QueryRow(hasEventsQuery, id).Scan(&hasEvents); err != nil {
+		return nil, nil, err
+	} else if !hasEvents {
+		return nil, nil, nil
+	}
+
+	const eventsQuery = `SELECT ev.id, ev.event_id, ev.maturity_height, ev.date_created, ci.height, ci.block_id, ev.event_type, ev.event_data
 	FROM events ev INDEXED BY events_maturity_height_id_idx -- force the index to prevent temp-btree sorts
+	INNER JOIN event_addresses ea ON (ev.id = ea.event_id)
```

Comment:
Tbh I'm quite surprised that after switching this to an `INNER JOIN` the previous query still makes that much of a difference. But I'm sure you have tested that combination of changes as well.

You'd expect that the db would immediately realise that there is nothing to join over and then return immediately. After all the first query performs the same join to figure out whether there are events at all.

Reply:
I think `INDEXED BY events_maturity_height_id_idx` is throwing off the query planner. Without that though it tries to do a temp b-tree sort on events, which is a massive table in full index mode.